### PR TITLE
bin: allow custom test path to be used

### PR DESCRIPTION
**Pull request by gdams**
This PR allows you to pass in `citgm-all --tmpDir /path/to/directory`

### CODE REVIEW

**Comment by gibfahn** [comment]
Does this work with `citgm.js` as well? You need to document this in the README.

Also I think this should also set the npm tmp directory if the env var isn't already set. I can't see people using this other than to make sure citgm isn't writing things to random folders in their system and filling up `/tmp`.

**Comment by gdams** [comment]
@gibfahn yes this works on citgm.js, README updated and not sure about 3

**Comment by gibfahn** [comment]
@MylesBorins thoughts?

>Also I think this should also set the npm tmp directory if the env var isn't already set. I can't see people using this other than to make sure citgm isn't writing things to random folders in their system and filling up /tmp.

I think we want less configuration rather than more.

File: lib/create-options.js [hunk 1]
```diff
@@ -9,7 +9,12 @@ function createOptions(cwd, context) {
   }
   options.env = Object.create(process.env);
   options.env['npm_loglevel'] = 'error';
-  options.env['npm_config_tmp'] = context.npmConfigTmp;
+  // Use the same directory as --tmpDir for npm
```
Comment (gibfahn) [inline lib/create-options.js]: +`if npm_config_tmp not set`

File: lib/temp-directory.js [hunk 2]
```diff
@@ -11,10 +11,20 @@ function create(context, next) {
   var _path = path.join(osenv.tmpdir(),uuid.v4());
   var npmConfigTmp = path.join(_path, 'npm_config_tmp');
 
-  context.path = _path;
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, context.module.name);
+    context.emit('data', 'verbose', context.module.name + ' mk.tempdir', context.path);
+    mkdirp(context.path, function (err) {
```
Comment (MylesBorins) [inline lib/temp-directory.js]: this async call is not handled correctly. I could potentially end up calling the callback twice.

File: lib/create-options.js [hunk 3]
```diff
@@ -9,7 +9,16 @@ function createOptions(cwd, context) {
   }
   options.env = Object.create(process.env);
   options.env['npm_loglevel'] = 'error';
-  options.env['npm_config_tmp'] = context.npmConfigTmp;
+  // Overwrite the npm temp directory if not already set via the env variable
+  if (!options.env['npm_config_tmp']) {
```
Comment (MylesBorins) [inline lib/create-options.js]: all the logic for the should be in `temp-directory.js` there should be no need for any checks or alternative code paths at this point

File: lib/temp-directory.js (hunk 2 above)
Comment (MylesBorins) [inline lib/temp-directory.js]: it is worth mentioning that `mkdirp` will make all directories to create the final directory. As such `mkdirp` of the `npm temp directory` will result in all other temp directories being made. This mkdirp is not neccessary.

File: test/test-create-options.js [hunk 4]
```diff
@@ -12,13 +12,16 @@ test('create-options:', function (t) {
     module: {envVar: {testenvVar: 'thisisatest'}}
   };
 
+  //ensure that there is no environment variable defined
+  delete process.env['npm_config_tmp'];
```
Comment (MylesBorins) [inline test/test-create-options.js]: I would prefer this to be set to null rather than delete. /cc @jasnell for an opinion

**Comment by gdams** [comment]
@MylesBorins @gibfahn updated PTAL

**Comment by gdams** [comment]
Kicking off a CI run to ensure that the problem has been fixed

File: test/test-create-options.js (hunk 4 above)
Comment (gdams) [inline test/test-create-options.js]: setting to null doesn't seem to work, npm by default sets `npm_config_tmp` when running `npm test` so we have to unset this to check that citgm can run without that environment variabale set

File: lib/npm/install.js [hunk 5]
```diff
@@ -18,6 +18,8 @@ function install(context, next) {
   var bailed = false;
   context.emit('data', 'info', context.module.name + ' npm:', 'npm install started');
 
+  context.emit('data', 'verbose', context.module.name + ' npm:', 'Using temp directory: "' + options.env['npm_config_tmp'] + '"');
```
Comment (gibfahn) [inline lib/npm/install.js]: Split long line:

```js
  context.emit('data', 'verbose', context.module.name + ' npm:',
      'Using temp directory: "' + options.env['npm_config_tmp'] + '"');
```

File: lib/temp-directory.js [hunk 6]
```diff
@@ -9,12 +9,32 @@ var osenv = require('osenv');
 
 function create(context, next) {
   var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
-
-  context.path = _path;
+  var npmConfigTmp = _path;
+  // Overwrite the npm temp directory if not already set via the env variable
+  if (!process.env.npm_config_tmp) {
+    // Use the same directory as --tmpDir for npm if set
```
Comment (gibfahn) [inline lib/temp-directory.js]: `--tmpDir`-> `the citgm temp directory` to make it clear that tmpDir is the Citgm temp directory

File: lib/temp-directory.js [hunk 7]
```diff
@@ -9,12 +9,32 @@ var osenv = require('osenv');
...
+    if (context.options && context.options.tmpDir) {
+      npmConfigTmp = path.join(context.options.tmpDir, 'npm_config_tmp');
+    } else {
+      // Else use the system tmp directory
```
Comment (gibfahn) [inline lib/temp-directory.js]: `tmp` -> `temp` for consistency

File: lib/temp-directory.js [hunk 8]
```diff
@@ -9,12 +9,32 @@ var osenv = require('osenv');
...
+      npmConfigTmp = path.join(_path, 'npm_config_tmp');
+    }
+  } else {
+    npmConfigTmp = process.env.npm_config_tmp;
```
Comment (gibfahn) [inline lib/temp-directory.js]: Add comment, maybe:

`// Use the env var's temp directory if set`

File: test/test-temp-directory.js [hunk 9]
```diff
@@ -23,6 +23,9 @@ var badContext = {
   }
 };
 
+//ensure that there is no environment variable defined
+delete process.env['npm_config_tmp'];
```
Comment (gibfahn) [inline test/test-temp-directory.js]: Maybe explain more:

```js
// When run through `npm test`, npm will set npm_config_tmp.
// Unset it to test the no env variable case.
```

File: lib/common-args.js [hunk 10]
```diff
@@ -31,6 +31,10 @@ module.exports = function commonArgs (app) {
     type: 'string',
     description: 'Path to the node source to use when compiling native addons'
   })
+  .option('tmpDir', {
+    type: 'string',
+    description: 'Directory to test modules in'
```
Comment (MylesBorins) [inline lib/common-args.js]: should we maybe include information about the directory being cleaned up?

File: lib/temp-directory.js [hunk 11]
```diff
@@ -2,21 +2,38 @@
...
 var rimraf = require('rimraf');
 var uuid = require('uuid');
 var osenv = require('osenv');
 
 function create(context, next) {
   var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
-
-  context.path = _path;
+  var npmConfigTmp = _path;
```
Comment (MylesBorins) [inline lib/temp-directory.js]: no need to set this, it is potentially confusing. perhaps just leave it as `var npmConfigTmp`

File: lib/temp-directory.js [hunk 12]
```diff
@@ -2,21 +2,38 @@
...
+    // Use the env var's temp directory if set
+    npmConfigTmp = process.env.npm_config_tmp;
+  }
   context.npmConfigTmp = npmConfigTmp;
 
-  context.emit('data', 'silly', context.module.name + ' mk.tempdir', _path);
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, context.module.name);
+  } else {
+    context.path = npmConfigTmp;
```
Comment (MylesBorins) [inline lib/temp-directory.js]: why would we set the context.path to npmConfigTmp? we want it to be _path, not the same directory

File: lib/temp-directory.js [hunk 13]
```diff
@@ -2,21 +2,38 @@
...
+  }
+  context.emit('data', 'verbose', context.module.name + ' mk.tempdir', context.path);
 
-  mkdirp(npmConfigTmp, function (err) {
+  async.map([context.path, npmConfigTmp], mkdirp, function(err) {
```
Comment (MylesBorins) [inline lib/temp-directory.js]: we should not use async if npmConfigTmp is inside of context.path

File: test/test-temp-directory.js [hunk 14]
```diff
@@ -23,6 +23,10 @@ var badContext = {
   }
 };
 
+// When run through `npm test`, npm will set npm_config_tmp.
+// Unset it to test the no env variable case.
+delete process.env['npm_config_tmp'];
```
Comment (MylesBorins) [inline test/test-temp-directory.js]: would it make sense to do this as a test? Should we cache the value in setup and reset in teardown. I'm not a big fan of having the tests leave the system different than when it found it

File: lib/temp-directory.js [hunk 15]
```diff
@@ -9,12 +9,18 @@ var osenv = require('osenv');
 
 function create(context, next) {
   var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
+  var npmConfigTmp = '';
```
Comment (MylesBorins) [inline lib/temp-directory.js]: this can simply be `var npmConfigTmp;`

File: lib/temp-directory.js [hunk 16]
```diff
@@ -9,12 +9,18 @@ var osenv = require('osenv');
...
 
-  context.path = _path;
-  context.npmConfigTmp = npmConfigTmp;
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, context.module.name);
+    npmConfigTmp = path.join(context.path, 'npm_config_tmp');
+  } else {
+    npmConfigTmp = path.join(_path, 'npm_config_tmp');
+    context.path = _path;
+  }
```
Comment (MylesBorins) [inline lib/temp-directory.js]: this whole block can be simplified.

```js
if (context.options && context.options.tmpdir) {
  context.path = path.join(context.options.tmpDir, context.module.name);
}
else {
  context.path = _path;
}
context.npmConfigTmp = path.join(context.path, 'npm_config_tmp');
```

Comment (MylesBorins) [inline lib/temp-directory.js]: should we be using `context.module.name`? Is is guaranteed to exist? Should we include citgm in here?

Comment (MylesBorins) [inline lib/temp-directory.js]: Also, with a custom path, we should likely still be using a UUID of some sort to make the folder unique

File: lib/temp-directory.js [hunk 17]
```diff
@@ -9,14 +9,17 @@ var osenv = require('osenv');
 
 function create(context, next) {
   var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
 
-  context.path = _path;
-  context.npmConfigTmp = npmConfigTmp;
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, context.module.name);
```
Comment (MylesBorins) [inline lib/temp-directory.js]: this line should still use uuid to generate a nonce. 

I think we can likely get rid of _path and make the path in line

File: test/test-temp-directory.js [hunk 18]
```diff
@@ -36,6 +47,18 @@ test('tempDirectory.create:', function (t) {
   });
 });
 
+test('tempDirectory.create --tmpDir:', function (t) {
+  tempDirectory.create(contextTmpDir, function (e, ctx) {
+    t.error(e);
+    t.equals(ctx.path, 'thisisatest/test-module', 'context should now have a path');
```
Comment (gdams) [inline test/test-temp-directory.js]: @MylesBorins `this line should still use uuid to generate a nonce.`
We will have to use some regex to match here if you want to use uuid.

**Comment by gdams** [comment]
@MylesBorins updated PTAL

File: lib/temp-directory.js [hunk 19]
```diff
@@ -8,15 +8,16 @@ var uuid = require('uuid');
...
 function create(context, next) {
-  var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
-
-  context.path = _path;
-  context.npmConfigTmp = npmConfigTmp;
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, uuid.v4(), context.module.name);
+  } else {
+    context.path = path.join(osenv.tmpdir(),uuid.v4());
```
Comment (MylesBorins) [inline lib/temp-directory.js]: missing space between tmpdir and uuid 

want to put together a pr with a lint rule to enforce this after this lands?

Comment (MylesBorins) [inline lib/temp-directory.js]: also worth mentioning that the module name should likely be included in both potential paths

File: lib/temp-directory.js [hunk 20]
```diff
@@ -8,15 +8,16 @@ var uuid = require('uuid');
 var osenv = require('osenv');
 
 function create(context, next) {
-  var _path = path.join(osenv.tmpdir(),uuid.v4());
-  var npmConfigTmp = path.join(_path, 'npm_config_tmp');
-
-  context.path = _path;
-  context.npmConfigTmp = npmConfigTmp;
+  if (context.options && context.options.tmpDir) {
+    context.path = path.join(context.options.tmpDir, uuid.v4(), context.module.name);
```
Comment (MylesBorins) [inline lib/temp-directory.js]: module name should not be an extra folder depth... uuid. 4() + context.module.name

yiu could potentially cache this above the if statement and use for both potential results

**Pull request merged**